### ISSUE TITLE
UART write issue

### ISSUE THREAD

**Issue by khavernathy** [MaximIntegratedAI/MaximAI_Documentation#73]
Hey folks, I have a strange error when trying to replicate what the `ImgCapture` example does with some of my own code. In `ImgCapture/main.c` the image raw byte data is written over UART for the `console.py` code to read. This is done either all-at-once or streamed in chunks depending on if CNN SRAM is used for the image data.

I'm finding in both cases (`transmit_capture_uart()` or `transmit_stream_uart()`), for a 128x128 image, about 90% of the bytes (+/- a few on repeat testing) are written over UART, instead of the whole 65536-byte chunk. This is an example of my terminal output (`console.py` feed)

```
MCU: W:128 H:128 L:65536
MCU: *IMG* RGB888 65536 128 128
Collecting 65536 bytes...
Image receive failed!  Only received 58366/65536 bytes
MCU: b'\xb8\x00\xe8\xec\xb8\x00\xe8\xec\xb8\x00\xe8\xec\xb8\x00\xe8\xec\xb8\x00\xe8\xf0\xb0\x00\xe8\xf0\xb0\x00\xe8\xec\xb8\x00\xe8\xec\xb8\x00\xe8\xe8\xb8\x00\xe8\xec\xb8\x00\xe8\xec\xb0\x00\xf0\xf0\xb8\x00\xe8\xec\xb0\x00\xf0\xf0\xb0\x00\xe8\xec\xb0\x00\xe8\xec\xb0\x00\xe8\xf0\xc0\x00\xf0\xf0\xc0\x00\xf0\xf0\xb8\x00\xf0\xf0\xb8\x00\xf0\xf0\xb8\x00\xf0\xf0\xb8\x00\xf0\xf4
.... (and a bunch more bytes)
```

`console.py`'s `Serial` instance is receiving/reading 58,366 bytes in this case, **and then the remaining 7,170 (exactly summing to desired 65,536) are printed onto the terminal** but evidently fail to send over UART.

I'm not sure what's going on here. I've tinkered with the baud rate (making sure python's `Serial` is listening at the same rate as specified in `console.h`

Am I running out of memory on the MAX78000? (I am storing some CNN weights and model logic in this program) Is there an easy way to monitor memory use? Any help appreciated.

**Comment by Jake-Carter** [comment]
Hi @khavernathy, I'll look into this.  Would you mind zipping up and attaching your project?

Yes, sometimes this is a symptom of memory issues.  You can check the `.bss` section in the build output, which will show how much static memory your application allocates.  (I like to remember as "better save space").  Dynamic memory is a bit more difficult to measure, so make sure you're error checking the return values of the camera init/setup functions.  They will pass through any error codes from the internal image buffer `malloc`s.

It could be other things though, possibly also on the Python side...  It's possible that the Python script is timing out during the image transfer.  The default is 5s, which should be enough for a 128x128 transfer at 921600 baud...  Try increasing the timeout to 10s and see if the transfer still fails.

```shell
~/repos/msdk/Examples/MAX78000/ImgCapture (main*) » python utils/console.py --help                    jhcarter@jhcarter-X670E
usage: console.py [-h] [--baudrate BAUDRATE] [--timeout TIMEOUT] port

positional arguments:
  port                 Serial port to connect to

options:
  -h, --help           show this help message and exit
  --baudrate BAUDRATE  Baud rate (default 921600)
  --timeout TIMEOUT    Communication timeout in seconds (default 5)
```

```shell
python utils/console.py --timeout 10 ...
```

FYI I'll also transfer this ticket to the [MSDK repo](https://github.com/Analog-Devices-MSDK/msdk)

**Comment by Jake-Carter** [comment]
Continue on https://github.com/Analog-Devices-MSDK/msdk/issues/816

**Comment by khavernathy** [comment]
Thanks very much.

Upping the timeout seems to have been my fix. I was slowing down the baudrate so i could push both regular printf output, and the images through the console.

I'm also finding tinkering with the camera frequency (e.g. 10e6 vs. 5e6) seems to make a difference in image quality.

**Comment by Jake-Carter** [comment]
> I'm also finding tinkering with the camera frequency (e.g. 10e6 vs. 5e6) seems to make a difference in image quality.

Yes, the exposure time and data-rate for these low-power CMOS sensors are proportional to the input clock.  The sampling of the image array is directly driven by the clock input, so by slowing it down you increase the exposure time.  The AGC/AEC will also be slowed down, but will still be functional and attempt to compensate for the increased exposure.  So by reducing the clock speed you're leaning more heavily into the correction engines in the digital core, which are also running a bit slower.

FYI if you'd like to increase the baud rate for `printfs` you can change `CONSOLE_BAUD` in your board.h file [here](https://github.com/Jake-Carter/msdk/blob/d6aaa2274c278b6de8dd847471c801def53905af/Libraries/Boards/MAX78000/FTHR_RevA/Include/board.h#L61).  Just clean & rebuild the project, and the console will re-initialize with the new speed [here](https://github.com/Jake-Carter/msdk/blob/d6aaa2274c278b6de8dd847471c801def53905af/Libraries/Boards/MAX78000/FTHR_RevA/Source/board.c#L108).  

If you're really interested in how this is linked to `printf` all the way down see our stdio implementation [here](https://github.com/Jake-Carter/msdk/blob/dev/cam02_reva/Libraries/MiscDrivers/stdio.c)

**Comment by Jake-Carter** [comment]
Happy to help!

**Comment by Jake-Carter** [comment]
Our camera drivers can be found in [`Libraries/MiscDrivers/Camera`](https://github.com/Analog-Devices-MSDK/msdk/tree/main/Libraries/MiscDrivers/Camera)

The OV7692 on the FTHR board is small and inexpensive, but not that great of a camera.

The Himax HM0360-AWA (see hm0360_color) is a decent alternative, but it's a raw bayer sensor without an ISP.  So there is some additional overhead required to debayer the images in the Arm core, which is not ideal.  Availability is also extremely limited for large quantities with long lead times.

I would recommend the GalaxyCore GC0308.  I'm finishing the software and adding support to the MSDK.  I would highly recommend this sensor.  The image quality is greatly improved, the AEC and white-balancing engines are very good, and the registers actually do what you expect once you decipher the somewhat limited documentation.  I have ported the facedetection models and developed a QR code detector with this camera the model performance is greatly improved over using the OV7692.

Shoot me an email at jake.carter@analog.com.  We have a new camera module reference design that's in engineering samples phase, and I can most likely get you samples/schematics.

![cam02_orientation](https://github.com/MaximIntegratedAI/MaximAI_Documentation/assets/38844790/8f8a55a7-827d-4e87-9f15-17e6140d5fcd)
